Add Visual Prompt Adapter in lightning side

harimkang opened the pull request:
### Summary

- Add Visual Prompt modules in lightning adapter
- Refactor lightning adapter (+anomalib adapter)
- Enable device option into Trainer (accelerator & devices)
    - default: "auto" / option: ("cpu", "gpu", "tpu", "ipu", "hpu", "mps")
- Unit-Test & Integration Test

### How to test
```bash
# Training
>>> otx train --data.task visual_prompting --data.train_data_roots tests/assets/car_tree_bug --data.val_data_roots tests/assets/car_tree_bug

# Unit-Test
>>> python -m pytest tests/v2/unit/adapters/torch/lightning
```

### Checklist

<!-- Put an 'x' in all the boxes that apply -->

- [x] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [x] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

harimkang committed:
Add prototype code for the new OTX API (#2419)

*Initialize prototype commits for new OTX APIs

harimkang committed:
Modify the Cython-module path links in setup.py

harimkang committed:
Write demo code for a model in mmpretrain

harimkang committed:
Add AutoEngine example for various mmpretrain model demo

harimkang committed:
Rename AutoEngine to AutoRunner

harimkang committed:
Modify mmpretrain engine.predict for multi-modal inference

harimkang committed:
Modify the absolute path of the dataset in the demo to a relative path

harimkang committed:
Remove breakpoint()

harimkang committed:
Merge branch 'develop' of https://github.com/openvinotoolkit/training_extensions into feature/v2

harimkang committed:
Add otx install stuff

harimkang committed:
Modify otx doctor command and enable rich-argparse helpformat

harimkang committed:
Fix CUDA_HOME default path & exist check

harimkang committed:
Fix mypy issues

harimkang committed:
Bumped torch up to 2.0.0

harimkang committed:
Add anomalib entry points

harimkang committed:
Fix precommit features

harimkang committed:
Fix Auto-Runner & cli with anomalib

harimkang committed:
Add TODO comments in cli.py

harimkang committed:
Merge branch 'develop' of https://github.com/openvinotoolkit/training_extensions into feature/v2

harimkang committed:
Add debug mode & fix importlib debug structure

harimkang committed:
Remove requirements/v2/ & Edit pyproject.toml with PEP621

harimkang committed:
Add AutoRunner entry points & Add demo for AutoRunner & Refactor BaseDataset

harimkang committed:
Update otx doctor table outputs & --debug to --verbose

harimkang committed:
Align classification outputs

harimkang committed:
Add -v short-term option command with otx doctor

harimkang committed:
Fix install string strip issue & Modify cli demo with new output paths

harimkang committed:
Enable rich in otx doctor traceback print

harimkang committed:
Add verbosity-level help format into cli subcommands

harimkang committed:
Enable Darwin platform install & train

harimkang committed:
Fix wrong optim_wrapper argument.

harimkang committed:
Enable mypy otx/v2/api, otx/v2/cli

harimkang committed:
Enable mypy otx/v2/adapters

harimkang committed:
Add feature to print Guide & TraceBack when exception occurs

harimkang committed:
Merge branch 'develop' of https://github.com/openvinotoolkit/training_extensions into feature/v2

harimkang committed:
Fix save config command

harimkang committed:
Fix pylint & pyflakes

harimkang committed:
Enable flake8-annotations

harimkang committed:
Enable flake8-annotations with adapters

harimkang committed:
Enable all ruff-stuff except not-founded & pydocstring

harimkang committed:
Add ruff settings

harimkang committed:
Fix pre-commit except adapters

harimkang committed:
Fix pre-commit except adapters.modules

harimkang committed:
Edit pyproject.toml

harimkang committed:
Add api core unit-tests

harimkang committed:
Move entity related modules into entites folder & add api utils unit-tests

harimkang committed:
Add api utils unit-test code

harimkang committed:
Copy & Fix adapters.datumaro test codes from 1.X

harimkang committed:
Copy & Fix adapters.openvino test cases from 1.X

harimkang committed:
Add cli/cli.py unit tests & Fix some bugs

harimkang committed:
Add cli unit-tests

harimkang committed:
Enable pydocstyle google convention

harimkang committed:
Fix some conflict with jsonargparse

harimkang committed:
Refactor CLI help formatter side

harimkang committed:
Copy torch.modules unit test from otx 1.X

harimkang committed:
Copy mmdeploy unit test from otx 1.X

harimkang committed:
Add mmpretrain engine unit-test

harimkang committed:
Add mmpretrain model api unit-test

harimkang committed:
Fix some wrong migration in classifier & head side, Add dataset unit-test

harimkang committed:
Fix some bug in NonLinearHead and Add mmpretrain integration tests

harimkang committed:
Remove # type: ignore

harimkang committed:
Remove # pylint: ignore comments

harimkang committed:
Fix Dataset adapter's use_mask

harimkang committed:
Remove # noqa as many as possible

harimkang committed:
Remove # pylint ignore in openvino adapters

harimkang committed:
Add integration test for auto-runner & anoamly_classification

harimkang committed:
Update ruff version

harimkang committed:
Update ruff & Enable all rules

harimkang committed:
Upgrade jsonargparse to 4.25.0 & Fix bugs in cli

harimkang committed:
Add cli integration test

harimkang committed:
Remove unnecessary torch import

harimkang committed:
Remove wrong added file

harimkang committed:
Fix some auto_runner variable

harimkang committed:
Fix some docstring & rename config file

harimkang committed:
Add verbose into install args

harimkang committed:
Add console print for complete of installation

harimkang committed:
Update latest code

harimkang committed:
Update mmpretrain modules codes

harimkang committed:
Pull otx develop branch

harimkang committed:
Rebase develop branch & Update v2 modules code

harimkang committed:
Fixed a bug that prevented data_roots from being read properly

harimkang committed:
Enable AdaptiveRepeatDataHook to classification

harimkang committed:
Fix dataloader update in hooks

harimkang committed:
Enable CI build with feature/v2 & Fix _update_config function bug (#2570)

* Update CODEOWNERS for v2

* Fix wrong argument in config (Hot-Fix)

* Disable feature/v2 with pre_merge & Create v2_pre_merge.yaml

harimkang committed:
Convert anomalib to lightning.anomalib

harimkang committed:
Remove Unnecessary torch imports

harimkang committed:
Initial code implementation for visual prompt task

harimkang committed:
Split torch.anomalib into torch.lightning and torch.lightning.anomalib [v2] (#2578)

* Convert anomalib to lightning.anomalib

harimkang committed:
Rebase latest commit

harimkang committed:
Add unit & integration test of visual_prompt Engine

harimkang committed:
Fix some bugs in lightning side

harimkang committed:
Upgrade torch version to 2.0.1 & pre-commit in v2 (#2581)

Upgrade torch version to 2.0.1 & pre-commit

harimkang committed:
Fix data format type in Visual Prompt Dataset

harimkang committed:
Merge branch 'feature/v2' of https://github.com/openvinotoolkit/training_extensions into harimkan/visual_prompt

harimkang committed:
Revert pipeline type

harimkang commented on src/otx/v2/adapters/torch/lightning/anomalib/engine.py:
```diff
@@ -113,6 +117,15 @@ def _update_callbacks(
             ),
         ]
 
+    def _load_checkpoint(
```
@ashwinvaidya17 @samet-akcay 
I have a question about the load_checkpoint part.
I wrote a function called adapters/torch/lightning/engine.py::LightningEngine::load_checkpoint.
<img width="354" alt="image" src="https://github.com/openvinotoolkit/training_extensions/assets/38045080/c529ca52-04e6-4f20-83cc-13f6e88ad4a5">

However, I realized that this is a problem when predicting or exporting from anomalib Module.
I want this function to load the model when a checkpoint comes in as input, but do I need to do something different?

harimkang committed:
Add unit-tests for visual_prompt adapter

harimkang committed:
Refactor utils function structure

harimkang committed:
Add CODEOWNERS for visual_prompt

harimkang committed:
Edit CODEOWNERS for visual_prompt

harimkang committed:
Rebase squash commit

harimkang committed:
Add SAM test_step & Fix load_checkpoint step in lightning side

harimkang committed:
Fix integration export step

harimkang committed:
Enable device into Trainer

harimkang committed:
Remove device check

harimkang commented:
@ashwinvaidya17 We'll have to revisit the Anomalib-side adapters later. For this PR, can you go into more detail on the LightningEngine part?

goodsong81 commented on src/otx/v2/adapters/torch/lightning/anomalib/engine.py:
```diff
@@ -75,21 +75,25 @@ def _update_logger(
         Returns:
             list[pl.Callback] | pl.Callback | None: Updated callbacks.
         """
+        _ = mode
```
Inconsistent in type annotation vs doc
Nitpick: _ = mode <- looks ugly. Might be due to pre-commit?

harimkang replied:
That function has been removed.

harimkang commented on src/otx/v2/adapters/torch/lightning/engine.py:
```diff
@@ -137,17 +137,45 @@ def _update_logger(
             list[pl.Callback] | pl.Callback | None: Updated callbacks.
         """
 
+    def _load_checkpoint(
```
@ashwinvaidya17

ashwinvaidya17 replied:
From a quick glance I think we can drop `_load_checkpoint`. In Anomalib we use a `LoadModelCheckpoint` callback because we had some issues with loading the checkpoint files using lightning. The recommended method to load checkpoints in lightning is to pass the checkpoint path to the `fit`, `test`, `predict`, and `validate` method. So basically, `self.trainer.fit(model, datamodule, ckpt_path="/path/to/model.ckpt")`. I can see that's what you are doing in `fit` and `validate`. We can then just add a check something like
```python
def predict(self, ...):
    if checkpoint is None:
        checkpoint = self.latest_model.get("checkpoint")
    self.trainer.predict(model, ..., ckpt_path=checkpoint)
```
If checkpoint is None in `self.latest_model` then it won't break the execution. Optionally, we can also print a warning if both the parameter and the checkpoint path in self.latest_model is None.

And, since the trainer method takes care of calling `model.load_state_dict`, we can leave custom functionality to the model implementers.

As for your original question regarding loading the model for exporting and predicting, we can also do something like

```python
def export(self, model, checkpoint, ...):
    if checkpoint is None:
        checkpoint = self.latest_model.get("checkpoint")
    model = model.__class__.load_from_checkpoint(checkpoint)
```
I am not sure if it is a good programming practice to call `__class__` though. But here is where they mention how to load the checkpoint https://lightning.ai/docs/pytorch/stable/common/checkpointing_basic.html#lightningmodule-from-checkpoint

harimkang replied:
Actually, the current flow seems to work fine, but I'll test your suggestion and report back. Thanks for the nice comment!

harimkang replied:
I try to modify. However, the test, predict & export part is not working.
![image](https://github.com/openvinotoolkit/training_extensions/assets/38045080/999355b6-6e0d-4541-8d1f-73206d5c3afa)
(Note that the visual prompt works fine, but not with anomalib.)
So I think we should leave it as is for now.

ashwinvaidya17 replied:
Yeah, let's not try to use this with anomalib. Since it is independent library let's treat lightning engine separate and we can address this differently in anomalib engine.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/anomalib/dataset.py:
```diff
@@ -94,18 +93,18 @@ def __init__(
         self,
         subset: str,
-        pipeline: str | al.Compose | None = None,  # transform_config
+        pipeline: dict | list | None = None,  # transform_config
```
Not for this PR, but we could perhaps think about renaming `pipeline` to a more descriptive name. It is not clear to me that this is related to data transforms

harimkang replied:
Yes I agree. We'll need a separate PR for this.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/visual_prompt/__init__.py:
```diff
@@ -0,0 +1,24 @@
+try:
+    import os
+
+    os.environ["FEATURE_FLAGS_OTX_VISUAL_PROMPTING_TASKS"] = "1"
```
can you explain why this is really needed?

samet-akcay commented on src/otx/v2/adapters/torch/lightning/visual_prompt/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+AVAILABLE = True
+VERSION = None
+DEBUG = None
```
What is the motivation to have these ?

samet-akcay commented on src/otx/v2/adapters/torch/lightning/visual_prompt/dataset.py:
```diff
@@ -0,0 +1,171 @@
+"""OTX adapters.torch.lightning.visual_prompt.Dataset API."""
```
would this be more descriptive

samet-akcay commented on src/otx/v2/adapters/torch/lightning/visual_prompt/model.py:
```diff
@@ -0,0 +1,87 @@
+from .modules import MODELS
+
+MODEL_CONFIG_PATH = Path(get_otx_root_path()) / "v2/configs/visual_prompt/models"
+MODEL_CONFIGS = get_files_dict(MODEL_CONFIG_PATH)
```
not sure if we should have these here.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/visual_prompt/modules/datasets/dataset.py:
```diff
@@ -0,0 +1,392 @@
+        )
+
+
+class VisualPromptInferenceDataset(Dataset):
```
would it be possible to utilize datamodule's `predict_dataloader` instead of creating an additional inference dataset? Is this to follow MMx interface?

harimkang replied:
The reason I created this is because the current `OTXVisualPromptingDataModule` is based on a `DatasetEntity`.
Instead, only the predict part is rewritten to not use a `DatasetEntity` (`VisualPromptInferenceDataset`).
So currently, this is an OTXVisualPromptingDataModule that can't be reused without DatasetEntity, and if we want to utilize the DataModule, we'll need to create another new DataModule class.

So my suggestion would be to refactor `OTXVisualPromptingDataset`, `OTXVisualPromptingDataModule `and `VisualPromptInferenceDataset` at once in the future (with removing `DatasetEntity`), what do you think?

sungmanc commented on src/otx/v2/adapters/torch/lightning/engine.py:
```diff
@@ -200,17 +233,19 @@ def train(
         self.config["trainer"] = self.trainer_config
-        if optimizer is not None:
+
+        if isinstance(optimizer, Optimizer):
```
Just a question, what happens if `optimizer==None`?

harimkang replied:
This will look for the optimizer in the config and raise an error if it's not there.

sungmanc commented on src/otx/v2/adapters/torch/lightning/visual_prompt/dataset.py:
```diff
@@ -0,0 +1,171 @@
+        config = OmegaConf.load(filename=config) if isinstance(config, str) else DictConfig({})
+        config = config.get("dataset", config)
+        image_size = config.get("image_size", 1024)
+        normalize = config.get("normalize", {})
```
In the mmXdataset, we used `get_default_pipeline` function for this initialization. is there reason to select this way for the lightening case?

sungmanc replied:
We might use https://github.com/openvinotoolkit/training_extensions/pull/2583/files#diff-3d5379cce455de27aa201ecffd46130371bd4639d69c431a96aee05a887df4a2R59

harimkang replied:
Maybe Lightning will be a different way, I'll revisit the pipeline later.

sungmanc commented on src/otx/v2/adapters/torch/lightning/visual_prompt/engine.py:
```diff
@@ -0,0 +1,284 @@
+
+
+class VisualPromptEngine(LightningEngine):
+    """Anomalib engine using PyTorch and PyTorch Lightning."""
```
Some docstrings need to be updated. Anomalib -> VisualPrompt

harimkang replied:
Thank you. Fixed!

sungmanc commented on src/otx/v2/adapters/torch/lightning/visual_prompt/model.py:
```diff
@@ -0,0 +1,87 @@
+    checkpoint: str | None = None,
+    **kwargs,
+) -> torch.nn.Module:
+    """Return a torch.nn.Module object based on the provided model configuration or anomalib model api.
```
```suggestion
    """Return a torch.nn.Module object based on the provided model configuration or visual prompt model api.
```

sungmanc replied:
Need to check all files again :)

harimkang replied:
Thank you. Fixed all docstring word.

sungmanc commented on src/otx/v2/adapters/torch/modules/utils/io.py:
```diff
@@ -0,0 +1,109 @@
+"""Collection of dataset I/O functions."""
```
All functions implemented in this file are not related to the torch itself. how about moving the functions?

sungmanc replied:
Same for the src/otx/v2/adapters/torch/modules/utils/mask_to_bbox.py. Let's consider the refactoring of these modules at further PR

harimkang replied:
Yeah, I think we're going to need to have another discussion about structure sooner or later.

sungmanc reviewed (commented):
Left some comments. I couldn't do the detailed review for the test cases.

goodsong81 commented on src/otx/v2/adapters/torch/lightning/visual_prompt/modules/datasets/dataset.py:
```diff
@@ -0,0 +1,392 @@
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions
+# and limitations under the License.
```
Why don't you use the shorter version?
```suggestion
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

harimkang replied:
Thanks for the comment. Fixed!

goodsong81 commented on src/otx/v2/adapters/torch/modules/utils/mask_to_bbox.py:
```diff
@@ -67,3 +71,76 @@ def mask2bbox(mask: np.ndarray) -> list[list[int]]:
+    return cv2.drawContours(gt_mask, np.asarray([contour]), 0, 1, -1)
+
+
+def generate_bbox(
```
(Just a comment, not change request)
Function name is misleading. It's kind of "purturb_bbox()", right? @sungchul2 
If this should be located in adapters/torch/modules/utils, it would be better to be more clear or general. It looks like Visual Prompt task specific logic.

sungchul1 replied:
Yes, this function was to generate bboxes from given points with some perturbations when training.
Fitted box can be generated with `offset_bbox=0` as the default setting.
If it seems ambiguous, we can choose one of two: 
1. divide it into fitted version and perturbed version, and perturbed version can call fitted version
2. change function name like `generate_bbox_with_perturbation`

harimkang replied:
Which is better? I think I'll go with your recommendation. :)

sungchul1 replied:
I think 1 seems better, what do you think @goodsong81 ?

goodsong81 replied:
BTW, do we need fitted version? It will be just one-liner.
like return [max(0, x1), max(0, y1), min(w, x2), min(h, y2)]

harimkang replied:
This currently seems to always be used as offset_bbox=0, right?
<img width="655" alt="image" src="https://github.com/openvinotoolkit/training_extensions/assets/38045080/f4de70b4-2a0b-4dd5-ad3d-00fdcee8f6c3">

harimkang replied:
Anyway, Should it look like this?
```python
def generate_fitted_bbox(
    x1: int,
    y1: int,
    x2: int,
    y2: int,
    width: int,
    height: int,
) -> list[int]:
    """Generate bounding fitted box.

    Args:
        x1, y1, x2, y2 (int): Bounding box coordinates.
        width (int): Width of image.
        height (int): Height of image.

    Returns:
        list[int]: Generated bounding box.
    """
    return [
        max(0, x1),
        max(0, y1),
        min(width, x2),
        min(height, y2),
    ]


def generate_bbox_with_perturbation(
    x1: int,
    y1: int,
    x2: int,
    y2: int,
    width: int,
    height: int,
    offset_bbox: int = 0,
) -> list[int]:
    """Generate bounding box with perturbation.

    Args:
        x1, y1, x2, y2 (int): Bounding box coordinates.
        width (int): Width of image.
        height (int): Height of image.
        offset_bbox (int): Offset to apply to the bounding box, defaults to 0.

    Returns:
        list[int]: Generated bounding box.
    """

    def get_randomness(length: int) -> int:
        if offset_bbox == 0:
            return 0
        rng = np.random.default_rng()
        return int(rng.normal(0, min(int(length * 0.1), offset_bbox)))

    return generate_fitted_bbox(
        x1 + get_randomness(width),
        y1 + get_randomness(height),
        x2 + get_randomness(width),
        y2 + get_randomness(height),
        width,
        height,
    )
```

sungchul1 replied:
Ah, fitted version is already there :)

harimkang replied:
I did a simple function renaming (generate_bbox_with_perturbation) on this for now.

goodsong81 commented on tests/v2/integration/cli/test_cli.py:
```diff
@@ -34,9 +34,9 @@ def test_otx_cli(task: str, tmp_dir_path: Path) -> None:
+        "--data.val_data_roots", TASK_CONFIGURATION[task]["val_data_roots"],
         "--max_epochs", "1",
         "--seed", "1234",
-        "--deterministic", "True",
```
Specific reason for this?

harimkang replied:
I found a simple error with LightningModule (maybe it's only happening locally on my end).
```
RuntimeError: cumsum_cuda_kernel does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True)'. You can turn off determinism just for this operation, or 
you can use the 'warn_only=True' option, if that's acceptable for your application. You can also file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding 
deterministic support for this operation.
```

With `export CUBLAS_WORKSPACE_CONFIG=:16:8`, this also works fine with no errors. I haven't looked closely yet, maybe I should leave this as True and look into the issue a bit more?
Or maybe I can revisit this at a later date.

sungchul1 replied:
https://github.com/openvinotoolkit/training_extensions/blob/268aed7a92b829d68ff1eb9093cc8f180c1d4dfa/src/otx/algorithms/visual_prompting/tasks/train.py#L68
I think it can be a simple solution, what do you think?

harimkang replied:
I added Sungchul's suggestion and reverted this code.

goodsong81 reviewed (commented):
Thank you for the work. Generally looks OK to me.
A few minor comments are:

sungchul1 commented on src/otx/v2/adapters/torch/lightning/visual_prompt/dataset.py:
```diff
@@ -0,0 +1,171 @@
+
+
+@add_subset_dataloader(SUBSET_LIST)
+class VisualPromptDataset(LightningDataset):
+    """A class representing a dataset for VisualPrompt tasks."""
+
+    def __init__(
+        self,
+        task: TaskType | str | None = TaskType.VISUAL_PROMPTING,
```
Maybe I missed this part, task name is `VisualPrompt` or visual_prompt but type name in `TaskType` is `VISUAL_PROMPTING`. How about unifying them into either visual prompt or visual prompt"ing"?

sungchul1 replied:
In v1, the latter (viusal prompting) seems to be set.

harimkang replied:
Ah yes, a bit of a refactoring, and a word with visual_prompting.

sungchul1 commented on src/otx/v2/adapters/torch/lightning/visual_prompt/modules/datasets/dataset.py:
```diff
@@ -0,0 +1,392 @@
+        prompts = {
+            "original_size": image.shape,
+            "gt_masks": [],
+            "bboxes": [],
+            "points": [],
+            "labels": [],
```
Current visual prompting task only enables inference with prompts, such as bboxes and points, so they must be included in `prompts` dict.

harimkang replied:
What do you think I should be doing here? Or should I just leave it as is for now and check it out later?

sungchul1 replied:
When I think it again, leaving it as is would be better.
Fine-tuning uses an image and those prompts together, so that dataset for inference is unnecessary.
But zero-shot learning to be implemented later can use that dataset with a few modifications because zsl uses only an image.

harimkang replied:
Okay, I'm going to leave this as is. thanks!

sungchul1 reviewed (commented):
LGTM. I left some comments.

ashwinvaidya17 reviewed (commented):
From my initial look I feel we might not need another Engine for Visual Prompting tasks. And, I feel we need to polish the adapters design a bit. Let's get on a call for that to brainstorm ideas. But here are my thoughts to start the discussions.

Taking a high-level view first  here is what I think the ideal workflow should be.

```python
model = SAM()
datamodule = VisualPromptingDatamodule(...)
engine = LightningEngine()
engine.train(model, datamodule)
```

Basically, let's try to do the least amount of work from within the engine. Ideally it should just be an interface to make a single API convenient, and do the `otx->backend` conversion.
If we have this design then we can also have a backend agnostic configuration file based on `jsonargparse` format.

```yaml
model:
    class_path: otx.models.visual_prompting.SAM
    init_args:
        ...
data:
    class_path: otx.data.visual_prompting.VisualPromptingDatamodule
    init_args:
       ...
engine:
    ...
```

To achieve this I feel we should revisit the folder structure. This is what I have in mind.

```bash
/otx
├── adapters
│   ├── anomalib
│   │   ├── dataloader.py
│   │   └── engine.py
│   ├── lightning
│   │   ├── dataloader.py
│   │   └── engine.py
│   └── mmx
│       └── engine.py
├── data
│   ├── mvtec.py
│   └── visual_prompting.py
└── models
    ├── anomaly
    │   ├── padim.py
    │   └── stfpm.py
    └── visual_prompting
        ├── callbacks
        ├── data
        ├── logger
        └── model.py
```

From my understanding,  visual prompting is more of an algorithm implementation rather than a backend/library. If we call some of the methods from the model then we can leave the implementation details to the algorithm implementer. Like moving the `onnx/openvino` export code inside `model.export`.

```python
class Engine:
	...
	def export(...):
		# do some setup/preprocessing stuff
		model.export()
```

Currently, the Engine is responsible for adding all the required hooks. This is (from what I see) the main reason we require a separate engine for the visual prompting task. We can consider having a callbacks inside the model class that the engine can refer to. so instead of
```python
class VisualPromptingEngine(LightningEngine):
	def _update_callbacks(...)
		...
```
we can do
```python
class LightningEngine(Engine):
	def train(..., callbacks:list[Callbacks] = [])
		callbacks.extend(model.callbacks)
```

To achieve all these we can look into having a base `OTXModel` that specifies all the required methods.

```python
class BaseOTXModel(ABC):
	@abstractproperty
	def callbacks(self)->list[Callbacks]:
		...
	@abstractmethod
	def export(self, ...):
		...
```

To put this together the pseudo-code of the `LightningEngine` can be
```python
class LightningEngine(Engine):
	def __init__(self, ...):
		self.trainer: Trainer
		...
	def train(self, ...):
		...
		callbacks.extend(model.callbacks)
		self.trainer.fit(model, callbacks=callbacks)
	def export(self, ...):
		...
		model.export()
		
```
And the visual prompting task can be
```python
class SAMModel(BaseOTXModel, LightningModule):
	...
	def export(self, ...):
		...
		torch.onnx.export(
            model=self.image_encoder,
            args=tuple(dummy_inputs.values()),
            f=onnx_dir / "sam_encoder.onnx",
           ...
           )
         ...
     @property
     def callbacks(self)->list:
	     return [
	      ModelCheckpointCallback(),
	      LearningRateMonitor(),
	     ]
```

harimkang commented:
> ```

@ashwinvaidya17 
I like your idea very much, and the reason this didn't happen earlier is probably due to my lack of familiarity with Lightning.
But this is a design that needs to be checked on the Anomalib side as well. so for now, I think we can work on Anomalib separately. in this PR, I can work on some of your suggestions for the Visual Prompt first.

but,
Changing the folder structure (+dataset side) is a more general design proposal that we need to discuss more.
So for now, I think I can test the suggestions from Engine & Model for visual prompt. (If possible)
If this is something don't have enough time, I think we should keep it as is (merge first) and discuss refactoring it later.

samet-akcay commented:
>@ashwinvaidya17
I like your idea very much, and the reason this didn't happen earlier is probably due to my lack of familiarity with Lightning.
But this is a design that needs to be checked on the Anomalib side as well. so for now, I think we can work on Anomalib separately. in this PR, I can work on some of your suggestions for the Visual Prompt first.
but,
Changing the folder structure (+dataset side) is a more general design proposal that we need to discuss more.
So for now, I think I can test the suggestions from Engine & Model for visual prompt. (If possible)
If this is something don't have enough time, I think we should keep it as is (merge first) and discuss refactoring it later.

Thanks @harimkang and @ashwinvaidya17

harimkang committed:
Fix docstring

harimkang committed:
Refactor Engine & Model

harimkang committed:
Refactor visual_prompt structures

harimkang committed:
Fix docstring

harimkang committed:
Fix License

harimkang committed:
Refactoring for Lightning & Visual Prompt

Refactoring for Lightning & Visual Prompt

harimkang commented:
@goodsong81 @samet-akcay @ashwinvaidya17 @sungmanc 
I did a refactoring that incorporated Ashwin's suggestions as much as possible.
I removed the visual_prompt folder in adapters/torch/lightning and modified each piece of code to the appropriate location. I modified the export and callbacks to be functions and properties of the base model so that they can be overridden.

`src/otx/v2/adapters/torch/lightning/modules/models/base_model.py` -> Models that inherit from that class can use basic exports; each must declare callbacks.
```python
# src/otx/v2/adapters/torch/modules/models/base_model.py
class BaseOTXModel(ABC):
    @abstractproperty
    def callbacks(self) -> list:

    @abstractmethod
    def export(
        self,
        export_dir: str | Path,
        export_type: str = "OPENVINO",
        precision: str | int | None = None,
    ) -> dict:

# src/otx/v2/adapters/torch/lightning/modules/models/base_model.py
class BaseOTXLightningModel(BaseOTXModel):
    config: DictConfig
    device: torch.device

    def export(
        self,
        export_dir: str | Path,
        export_type: str = "OPENVINO",
        precision: _PRECISION_INPUT | None = None,
    ) -> dict:
        ... Common export code for LightningModule ...

```

Example:
- Anomaly Model: `src/otx/v2/adapters/torch/lightning/anomalib/model.py` -> Modified to wrap the model so that it can inherit from that class. This is temporary implementation and may change in future anomalib v1.0 updates and refactorings.
- SAM: `src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py` -> inherited from BaseModel and wrote an export function and callbacks.

```python
# src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py
class SegmentAnything(BaseOTXLightningModel, LightningModule):
    ...
    @abstractproperty
    def callbacks(self) -> list:
        ...

    def export(
        self,
        export_dir: str | Path,
        export_type: str = "OPENVINO",
        precision: _PRECISION_INPUT | None = None,
    ) -> dict:
       ... Override export function ...
```

LightningEngine: `src/otx/v2/adapters/torch/lightning/engine.py` -> Modified the using model.export, model.callbacks.

Limitations:
- We should consider a good way to clearly separate our Tasks. For now, we'll simply use str.
- Added code for Visual Prompting only to the LightningEngine.predict function. This is not generic and would be nice to clean up in the future. Any ideas are welcome.
- I haven't gone through all the suggestions yet (ex. configuration or folder structure). The other changes are overall design aspects that we'd like to see implemented after more discussion.

harimkang committed:
Fix conflict with new commit from feature/v2

harimkang committed:
Merge branch 'feature/v2' into harimkan/visual_prompt

harimkang committed:
Fix deterministic & Split generate_bbox function

harimkang committed:
Edit mask_to_bbox

harimkang committed:
Merge branch 'harimkan/visual_prompt' of https://github.com/harimkang/training_extensions into harimkan/visual_prompt

sungmanc reviewed (approved):
Thanks for the nice refactoring

harimkang committed:
Merge branch 'feature/v2' into harimkan/visual_prompt

samet-akcay commented on src/otx/v2/adapters/torch/lightning/__init__.py:
```diff
@@ -2,3 +2,24 @@
+    VERSION = pl.__version__
+    import os
+
+    os.environ["FEATURE_FLAGS_OTX_VISUAL_PROMPTING_TASKS"] = "1"
```
Can you explain in the code why this is needed?

harimkang replied:
This is an environment variable to prevent Geti from collecting the VISUAL_PROMPTING TaskType and its model. It may be cleared in a future Geti integration, but for now we need this code because it's tied into the TaskType import on the Entity side.

ashwinvaidya17 replied:
Maybe we can then add a TODO explaining this and mentioning when we need to remove this line

samet-akcay commented on src/otx/v2/adapters/torch/lightning/__init__.py:
```diff
@@ -2,3 +2,24 @@
 # Copyright (C) 2023 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
+
+AVAILABLE = True
+VERSION = None
+DEBUG = None
```
Can you also explain why these are needed, and where they are used?

harimkang replied:
This is used by AutoRunner and the CLI. Before importing everything, we can get information about the availability and version of these adapter, and if they are not available, we can save and check the exception. All adapters currently have this information.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/anomalib/engine.py:
```diff
@@ -128,7 +105,8 @@ def train(
         val_interval: int | None = None,
         logger: list[Logger] | Logger | None = None,
-        callbacks: list[pl.Callback] | pl.Callback | None = None,
+        callbacks: list | None = None,
```
does this mean that `callbacks` could be an object that is not `pl.Callback`?

harimkang replied:
I've seen some examples where we can use a callback without inheriting from pl.Callback.
But I don't know if it's available in OTX.
should we be more explicit and limit it to list[pl.Callback]?

harimkang replied:
I modified this to list[pl.Callback].

samet-akcay commented on src/otx/v2/adapters/torch/lightning/anomalib/engine.py:
```diff
@@ -167,25 +146,80 @@ def train(
+            callbacks (list[pl.Callback] | pl.Callback | DictConfig | None, optional): callbacks to use in test.
+            device (str | None, optional): Supports passing different accelerator types
+                ("cpu", "gpu", "tpu", "ipu", "hpu", "mps", "auto")
+            **kwargs: Additional keyword arguments to pass to the method.
```
Are these `kwargs` `lightning.Trainer` arguments?

harimkang replied:
Yes, that's right. I'll add a note that it's used in Trainer.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/anomalib/engine.py:
```diff
@@ -167,25 +146,80 @@ def train(
-        device: list | None = None,  # ["auto", "cpu", "gpu", "cuda"]
         logger: list[Logger] | Logger | None = None,
         callbacks: list[pl.Callback] | pl.Callback | None = None,
+        device: str | None = "auto",  # ["auto", "cpu", "gpu", "cuda"]
```
```suggestion
        device: str = "auto",  # ["auto", "cpu", "gpu", "cuda"]
```

harimkang replied:
Thanks for the suggestion. I will fix it.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/anomalib/engine.py:
```diff
@@ -167,25 +146,80 @@ def train(
             img (Optional[Union[PREDICT_FORMAT, LightningDataModule]]): The input data to run inference on.
             checkpoint (Optional[Union[str, Path]]): The path to the checkpoint file to use for inference.
-            device (Optional[list]): The device to use for inference. Can be "auto", "cpu", "gpu", or "cuda".
+            device (Optional[str]): The device to use for inference. Can be "auto", "cpu", "gpu", or "cuda".
```
Looks like these `Args` section requires an update in terms of the annotations. They are all using the old annotation format. For example `Optional[str]`

samet-akcay commented on src/otx/v2/adapters/torch/lightning/anomalib/engine.py:
```diff
@@ -205,16 +239,71 @@ def predict(
+
+        if model is not None and checkpoint is not None:
+            self._load_checkpoint(model, checkpoint)
+        # Lightning Inferencer
```
This comment seems redundant.

harimkang replied:
I'll remove it.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/anomalib/engine.py:
```diff
@@ -205,16 +239,71 @@ def predict(
+        model: BaseOTXLightningModel | pl.LightningModule | None = None,
+        checkpoint: str | Path | None = None,
+        precision: _PRECISION_INPUT | None = None,
+        export_type: str = "OPENVINO",  # "ONNX" or "OPENVINO"
```
`  # "ONNX" or "OPENVINO"` options are already defined in the docstring, so we don't need them here.
```suggestion
        export_type: str = "OPENVINO",
```

harimkang replied:
I'll fix it.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/anomalib/engine.py:
```diff
@@ -205,16 +239,71 @@ def predict(
+        checkpoint: str | Path | None = None,
+        precision: _PRECISION_INPUT | None = None,
+        export_type: str = "OPENVINO",  # "ONNX" or "OPENVINO"
+    ) -> dict:
```
is it possible to be more specific about the annotation? Is it `dict[str, Any]`, or maybe more specific? It would be good to know the return type by looking at the annotation without diving deeper into the code.

harimkang replied:
This is possible, but it can also have a deeper structure. 
```python
{
    "onnx": {"encoder": encoder_onnx_path, "sam": sam_onnx_path}, 
    "openvino": {
          "encoder": {"xml": encoder_xml_path, "bin": encoder_bin_path},
          "sam": {"xml": sam_xml_path, "bin": sam_bin_path}
     }
}
```
For now, how about making dict something like dict[str, str | dict]?

harimkang replied:
I modified this to dict[str, str | dict] for now.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/anomalib/model.py:
```diff
@@ -7,22 +7,68 @@
     import torch
 
-MODEL_CONFIG_PATH = Path(get_otx_root_path()) / "v2/configs/anomaly_classification/models"
+MODEL_CONFIG_PATH = Path(get_otx_root_path()) / "v2/configs/lightning/models"
```
Would this line work even when users install otx via PyPI?

harimkang replied:
This needs to be tested. 
@goodsong81 This is the folder with the model YAML file, what do we need to do to get this into the package?

samet-akcay commented on src/otx/v2/adapters/torch/lightning/dataset.py:
```diff
@@ -29,17 +35,95 @@ def _initialize(self) -> None:
+        if not self.initialize:
+            self._initialize()
+
+        # NOTE: It needs to be refactored in a more general way.
```
I agree with the note. This is not a good practice to include task specific implementations here

harimkang replied:
Yes I personally would like to see this considered along with the anomalib adapter refactoring.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/engine.py:
```diff
@@ -292,82 +332,106 @@ def test(
             checkpoint (Optional[Union[str, Path]]): The checkpoint to use for testing.
                 If not provided, the latest checkpoint will be used.
             precision (Optional[_PRECISION_INPUT]): The precision to use for testing.
+            logger (list[Logger] | Logger | bool | None, optional): Logger to use in test.
```
looks like the docstring still has the old annotations. See Line 332, for example.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/engine.py:
```diff
@@ -292,82 +332,106 @@ def test(
 
         Args:
-            model (Optional[Union[torch.nn.Module, pl.LightningModule]]): The model to use for inference.
+            model (Optional[Union[BaseOTXLightningModel, pl.LightningModule]]): The model to use for inference.
```
?
```suggestion
            model (BaseOTXLightningModel | pl.LightningModule | None, optional): The model to use for inference.
```

samet-akcay commented on src/otx/v2/adapters/torch/lightning/model.py:
```diff
@@ -0,0 +1,86 @@
+
+from .modules.models import MODELS
+
+MODEL_CONFIG_PATH = Path(get_otx_root_path()) / "v2/configs/lightning/models"
+MODEL_CONFIGS = get_files_dict(MODEL_CONFIG_PATH)
```
Again, not sure if this is a good idea to parse the configs from the repo, which probably would not work when users install the library via `pip install otx`

harimkang replied:
But we currently provide the model of OTX as a config file. Also, in my opinion, these files should be packaged (like mmX), what do you think?

samet-akcay commented on src/otx/v2/adapters/torch/lightning/model.py:
I think we need to re-visit this module

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/__init__.py:
```diff
@@ -0,0 +1,4 @@
+"""Lightning based Adapter Modules."""
```
```suggestion
"""Lightning-based Adapter Modules."""
```

harimkang replied:
Fixed.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/datasets/__init__.py:
```diff
@@ -0,0 +1,20 @@
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from .dataset import (
+    OTXVisualPromptingDataModule,
+    OTXVisualPromptingDataset,
+    get_transform,
+)
```
This does not fit into a single line?
```suggestion
from .dataset import OTXVisualPromptingDataModule, OTXVisualPromptingDataset, get_transform
```

harimkang replied:
Yes, I'll try that.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/datasets/dataset.py:
```diff
@@ -0,0 +1,380 @@
+from torch.utils.data import DataLoader, Dataset
+from torchvision import transforms
+
+from otx.v2.adapters.torch.modules.utils.io import (
+    get_image_filenames,
+    read_image,
+)
+from otx.v2.adapters.torch.modules.utils.mask_to_bbox import (
+    convert_polygon_to_mask,
+    generate_bbox_from_mask,
+)
```
haven't checked, but I have the feeling that these imports could fit into a single line, with a length of 120?

harimkang replied:
i'll try it.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/datasets/dataset.py:
```diff
@@ -0,0 +1,380 @@
+from otx.v2.api.entities.subset import Subset
+from otx.v2.api.utils.logger import get_logger
+
+from .pipelines import (
+    MultipleInputsCompose,
+    Pad,
+    ResizeLongestSide,
+    collate_fn,
+)
```
same as above, single line?

harimkang replied:
i'll try it.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/datasets/dataset.py:
```diff
@@ -0,0 +1,380 @@
+    )
+
+
+class OTXVisualPromptingDataset(Dataset):
```
I'm not reviewing any of these that use `DatasetEntity` as they will be removed in the future

harimkang replied:
That's my hope too.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/datasets/pipelines/sam_transforms.py:
```diff
@@ -0,0 +1,114 @@
+
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
```
Should we also provide the license type here?
Also if we are modifying any of the classes/functions in this file, I think we also need to add intel header too. @yunchu, can you comment here?

harimkang replied:
I took Yunchu's advice and modified it as shown below.
```
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
# This code is a modification of the https://github.com/facebookresearch/segment-anything
```

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/datasets/pipelines/sam_transforms.py:
```diff
@@ -0,0 +1,114 @@
+            item (dict[str, np.ndarray]): Dictionary of batch data.
+
+        Returns:
+        Dict[str, Union[List, Tensor]]: Dictionary of batch data.
```
```suggestion
            dict[str, np.ndarray | list]: Dictionary of batch data.
```

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/datasets/pipelines/transforms.py:
```diff
@@ -0,0 +1,117 @@
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
```
```suggestion
```

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/datasets/pipelines/transforms.py:
As @ashwinvaidya17 suggested in his proposal, I think this part should ideally go to the following directory in the future:
```otx.data.<foundation, or visual_prompting>.transforms``` or something similar

harimkang replied:
Yes, we should discuss this in the forum.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/backbones/vit.py:
```diff
@@ -0,0 +1,478 @@
+"""Vision Transformers."""
+
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
again, please check the license.

harimkang replied:
Commented at https://github.com/openvinotoolkit/training_extensions/pull/2583#discussion_r1382720732

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/backbones/tiny_vit.py:
```diff
@@ -0,0 +1,837 @@
+"""TinyViT for MobileSAM."""
+
+# Copyright (c) 2022 Microsoft
```
license

samet-akcay replied:
Did we double check that these backbones are not available in timm, or torchvision?

harimkang replied:
@yunchu What should we do about the licensing of these files?
@sungchul2 Could you answer the second one?

sungchul1 replied:
I checked vit_b implemented in otx (=official SAM), timm, and torchvision), and they seem to have so different architectures. Since it is not usable by simply matching the checkpoint names of the layers, it seems best to implement the model as it is now.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/backbones/vit.py:
```diff
@@ -0,0 +1,478 @@
+
+
+# This class and its supporting functions below lightly adapted from the ViTDet backbone available at: https://github.com/facebookresearch/detectron2/blob/main/detectron2/modeling/backbone/vit.py
+class ViT(nn.Module):
```
again, would it be possible to use torchvision or timm?, no?

sungchul1 replied:
Commented at the above (https://github.com/openvinotoolkit/training_extensions/pull/2583#discussion_r1382738088)

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/encoders/sam_image_encoder.py:
```diff
@@ -0,0 +1,45 @@
+            self.backbone = build_vit(config.backbone, config.image_size)
+        else:
+            msg = f"{config.backbone} for image encoder of SAM is not implemented yet. Use vit_b, l, or h."
+            raise NotImplementedError(
+                msg,
+            )
```
```suggestion
            raise NotImplementedError(msg)
```

harimkang replied:
thanks, i'll fix it

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/encoders/sam_prompt_encoder.py:
```diff
@@ -0,0 +1,272 @@
+
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
```
again, specific license type, and Intel header

harimkang replied:
Commented at https://github.com/openvinotoolkit/training_extensions/pull/2583#discussion_r1382720732

samet-akcay commented on src/otx/v2/adapters/torch/modules/models/__init__.py:
```diff
@@ -0,0 +1,4 @@
+"""Models for OTX Torch Adapters."""
+# Copyright (C) 2023 Intel Corporation
```
```suggestion

# Copyright (C) 2023 Intel Corporation
```

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py:
```diff
@@ -0,0 +1,831 @@
+            self.image_embedding_size = self.config.model.image_size // patch_size
+        else:
+            msg = f"{self.config.model.backbone} for image encoder of SAM is not implemented yet. Use vit_b, l, or h."
+            raise NotImplementedError(
+                msg,
+            )
```
```suggestion
            raise NotImplementedError(msg)
```

harimkang replied:
i'll fix it

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py:
```diff
@@ -0,0 +1,831 @@
+        """Load checkpoint for SAM.
+
+        Args:
+            state_dict (Optional[OrderedDict], optional): State dict of SAM. Defaults to None.
```
```suggestion
            state_dict (OrderedDict, optional): State dict of SAM. Defaults to None.
```

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py:
```diff
@@ -0,0 +1,831 @@
+
+        Args:
+            state_dict (Optional[OrderedDict], optional): State dict of SAM. Defaults to None.
+            revise_keys (List, optional): List of tuples of regex patterns to revise keys of state_dict.
```
```suggestion
            revise_keys (list, optional): List of tuples of regex patterns to revise keys of state_dict.
```

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py:
```diff
@@ -0,0 +1,831 @@
+
+    ##########################################################
+    #     forward for inference (export/deploy/optimize)     #
+    ##########################################################
```
do we need this comment?

harimkang replied:
Removed unnecessary comments.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py:
```diff
@@ -0,0 +1,831 @@
+        point_labels: Tensor,
+        mask_input: Tensor,
+        has_mask_input: Tensor,
+        # orig_size: Tensor,
```
left-over?

harimkang replied:
Removed unnecessary arg.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py:
```diff
@@ -0,0 +1,831 @@
+
+        return pred_masks, ious
+
+    def training_step(self, batch: dict, batch_idx: int) -> Tensor:
```
I think the following methods would require a refactor:

- `forward`
- `forward_train`
- `training_step`

Ideally all of the training computation should be done a single `SAMLoss(nn.Module)` class, which should return the overall loss. And there should be a single `forward` method that splits training and inference stages. 

Maybe we could add this as a todo here for now, by creating a Jira ticket and leaving it here.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py:
```diff
@@ -0,0 +1,831 @@
+
+        pred_masks, iou_predictions = self.forward_train(images, bboxes, points)
+
+        loss_dice = 0.0
```
As I mentioned above, this loss computation here does not seem to be Lightning friendly. It could potentially be a lot simpler if we have a loss component, which Lightning can automatically update.

samet-akcay replied:
For example, refer to this loss example 
https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/models/dsr/loss.py

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py:
```diff
@@ -0,0 +1,831 @@
+    def export(
+        self,
+        export_dir: str | Path,
+        export_type: str = "OPENVINO",
```
maybe we could have `export_type` as Enum-str in otx in the future.

harimkang replied:
yes, i agree.

samet-akcay commented on src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py:
```diff
@@ -0,0 +1,831 @@
+        Returns:
+            list[Callback]: A list of callbacks.
+        """
+        callback_list = [
+            TQDMProgressBar(),
+        ]
```
```suggestion
        callback_list = [TQDMProgressBar()]
```

harimkang replied:
i'll fix it

samet-akcay commented on tests/v2/unit/adapters/torch/lightning/modules/datasets/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
```
some license headers have this extra line `#`, while some don't. it would be good to have consistency

harimkang replied:
I fixed this for the files in that PR.

samet-akcay reviewed (changes requested):
Thanks @harimkang. It was a huge PR to work on and review. I've some comments, but some of them are for future reference.

harimkang reviewed (commented):
@samet-akcay Thanks for reviewing.
I've answered most of what I could.
There are a few comments related to docstrings that I'll fix and get back to you later.
A few comments were on old commits, so I skipped those, please note.

@yunchu There are some license messages in the files related to SAM, can you tell me if it's okay to leave them as they are?
@sungchul2 Some of Samet's comments are SAM-specific details that you may be able to answer.

harimkang committed:
Incorporated some review comments (1st commit)

harimkang committed:
Fix docstring & license

harimkang commented:
@samet-akcay I've incorporated docstring (& license) and some of the fixes into the PR; I've resolved the corresponding comments to distinguish between them.

harimkang committed:
Remove unnecessary comments

harimkang committed:
Fix license comments

ashwinvaidya17 commented on src/otx/v2/adapters/torch/lightning/modules/__init__.py:
```diff
@@ -0,0 +1,4 @@
+"""Lightning-based Adapter Modules."""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
```
Maybe for a different PR but we need to revisit the folder structure a bit. Currently the backbones, encoders, decoders, and the SAM model are in the same hierarchy. Maybe under models we can have components, utils, and the entire model. But ofc this refactor depends on our call so we can continue as is.

ashwinvaidya17 commented on src/otx/v2/adapters/torch/lightning/modules/datasets/dataset.py:
```diff
@@ -0,0 +1,370 @@
+logger = get_logger()
+
+
+def get_transform(
```
Should this be a private method? Do we refer to this anywhere else?

harimkang replied:
Currently, this is the method used by `OTXVisualPromptingDataset` and `VisualPromptInferenceDataset`, so i'll leave it as is for now. It should be refactored together when these datasets are modified in the future.

ashwinvaidya17 commented on src/otx/v2/adapters/torch/lightning/modules/datasets/dataset.py:
```diff
@@ -0,0 +1,370 @@
+"""Dataset & DataModule for OTX lightning modules."""
```
How about we rename it to something like `visual_prompting.py`? This way, we can have other `lightning->otx` datasets in the same folder.

harimkang replied:
Thanks for the comment.
I renamed this to `visual_prompting_dataset.py`.

ashwinvaidya17 commented on src/otx/v2/adapters/torch/lightning/modules/datasets/pipelines/sam_transforms.py:
```diff
@@ -0,0 +1,114 @@
+"""SAM transfrom pipeline."""
```
If this is only specific to sam then it should be located with rest of the sam code. But again, we haven't discussed the folder structure so this is just a comment for the future.

ashwinvaidya17 reviewed (approved):
I think it looks good for now. There are some things that can be refactored but I don't think we can do that before some design discussions.

samet-akcay reviewed (approved):
Thanks for going over this pr multiple iterations @harimkang. We need to re-visit this later, but it is good for now!

harimkang committed:
Rename lightning.modules.datasets.dataset to visual_prompting_dataset
